lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -428,7 +440,12 @@ def apply_all(resources, prune)
         command.push("-f", tmp_dir)
 
         if prune
-          command.push("--prune", "--all")
+          command.push("--prune")
+          if @selector
+            command.push("--selector", Utils.selector_to_string(@selector))
+          else
+            command.push("--all")
+          end
           prune_whitelist.each { |type| command.push("--prune-whitelist=#{type}") }
```

Comment:
@DazWorrall @KnVerey Is it reasonable to only apply the whitelist when a selector is not provided, in order to satisfy https://github.com/Shopify/kubernetes-deploy/issues/271? This approach gives the user full flexibility to prune any kind of resources they like, such that we don't have to keep updating the whitelist, it being only a safe default.

Alternatively, what if we begin to require a selector and otherwise prune all resources?

Reply:
> We don't currently prune any `apps/v1/*` resources, for example, to which the likes of `Deployment` have been promoted.

We do actually. Group versions are better thought of as "views" on resources. If a `Deployment` exists, it can be accessed via any version supported by the API server in that cluster. In fact, you _must_ specific a group version whenever you `get` a resource--kubectl hides it from you, but try the following for example: 

```
kubectl get deployment.v1beta1.apps -v=6
GET https://localhost:6443/apis/apps/v1beta1/namespaces/katrina-test/deployments?limit=500 200 OK

kubectl get deployment.v1beta1.extensions -v=6
GET https://localhost:6443/apis/extensions/v1beta1/namespaces/katrina-test/deployments?limit=500 200 OK
```

In any case, so as long as we specify a group version that all of our supported k8s versions recognize (which we have CI coverage to prove), we will prune resources of that kind regardless of the group version used to create/update them.

> or making the prune whitelist configurable

I feel strongly that we should not expose the prune whitelist to end users. Pruning is technically an alpha feature that has some unfortunate gotchas (like the one I explained above where the whitelist requires but ignores GVs), and there's no such thing as pruning with no whitelist: kubectl has one (a much less sensible one) baked in. We should manage (i.e. prune) all kinds we support with few or maybe even no exceptions, and we should do so consistently whether we're managing the whole namespace or a selector-based sub-section of it.

> such that we don't have to keep updating the whitelist, it being only a safe default.

I don't actually think it's a big deal to manage this. We should only need to change it when we add support for new resources, and in the extremely rare cases where the group version we're referencing for a kind is deprecated. https://github.com/Shopify/kubernetes-deploy/issues/271 is really easy to fix and reflects omissions on our part earlier on in the development of this gem.

> Is it reasonable to only apply the whitelist when a selector is not provided

Overall I'm not sure I understand your reasoning with this proposal. Since pruning and selectors are both kubectl features, I (perhaps naively) don't anticipate problems with their interaction. The outstanding problem with secrets from EJSON is that users currently have no way to get them labelled correctly for use with sub-namespace deploys. Perhaps we can do that automatically; otherwise, we need to make it possible in the ejson file itself.